core/peerset/src/lib.rs

```diff
@@ -103,130 +199,149 @@ pub struct PeersetConfig {
+		self.data.discovered.mark_not_reserved(&peer_id);
+		if self.data.reserved_only {
+			if self.data.in_slots.contains(&peer_id) || self.data.out_slots.contains(&peer_id) {
+				self.data.in_slots.clear_slot(&peer_id);
+				self.data.out_slots.clear_slot(&peer_id);
+				self.message_queue.push_back(Message::Drop(peer_id))
+			}
+		}
 	}
 
-	/// Sets whether or not the peerset only has connections .
-	pub fn set_reserved_only(&self, reserved_only: bool) {
-		let mut inner = self.inner.lock();
-		let inner = &mut *inner;	// Fixes a borrowing issue.
-		inner.reserved_only = reserved_only;
-
+	fn on_set_reserved_only(&mut self, reserved_only: bool) {
 		// Disconnect non-reserved nodes.
-		if reserved_only {
-			for slot in inner.out_slots.iter_mut().chain(inner.in_slots.iter_mut()) {
-				if let Some(peer) = slot.as_ref() {
-					if inner.reserved.contains(peer) {
-						continue;
-					}
-
-					let _ = self.tx.unbounded_send(Message::Drop(peer.clone()));
-				}
-
-				*slot = None;
+		self.data.reserved_only = reserved_only;
+		if self.data.reserved_only {
+			for peer_id in self.data.in_slots.clear_common_slots().into_iter().chain(self.data.out_slots.clear_common_slots().into_iter()) {
+				// peer will be removed from `in_slots` or `out_slots` in `on_dropped` method
+				self.data.in_slots.clear_slot(&peer_id);
+				self.message_queue.push_back(Message::Drop(peer_id.clone()));
+				self.data.discovered.add_peer(peer_id, SlotType::Common);
 			}
+		} else {
+			self.alloc_slots();
 		}
 	}
 
-	/// Reports an adjustement to the reputation of the given peer.
-	pub fn report_peer(&self, _peer_id: &PeerId, _score_diff: i32) {
-		// This is not implemented in this dummy implementation.
-	}
-}
+	fn on_report_peer(&mut self, peer_id: PeerId, score_diff: i32) {
+		let score = self.data.scores.entry(peer_id.clone()).or_default();
+		*score = score.saturating_add(score_diff);
 
-fn alloc_slots(inner: &mut Inner, tx: &mpsc::UnboundedSender<Message>) {
-	if inner.reserved_only {
-		return;
-	}
-
-	for slot in inner.out_slots.iter_mut() {
-		if slot.is_some() {
-			continue;
+		if *score < 0 {
+			// peer will be removed from `in_slots` or `out_slots` in `on_dropped` method
+			if self.data.in_slots.contains(&peer_id) || self.data.out_slots.contains(&peer_id) {
+				self.data.in_slots.clear_slot(&peer_id);
+				self.data.out_slots.clear_slot(&peer_id);
+				self.message_queue.push_back(Message::Drop(peer_id));
+			}
 		}
+	}
 
-		if !inner.discovered.is_empty() {
-			let elem = inner.discovered.remove(0);
-			*slot = Some(elem.clone());
-			let _ = tx.unbounded_send(Message::Connect(elem));
+	fn alloc_slots(&mut self) {
+		while let Some((peer_id, slot_type)) = self.data.discovered.pop_peer(self.data.reserved_only) {
+			match self.data.out_slots.add_peer(peer_id.clone(), slot_type) {
+				Ok(_) => {
+					self.message_queue.push_back(Message::Connect(peer_id));
+				},
+				Err(SlotError::AlreadyConnected(_)) => (),
```

Comment:
I'd print an error here, since there's an obvious problem somewhere.

Reply:
good point! being here means that we have a logic error somewhere